### ISSUE TITLE
random behavior is not able to be controlled

### ISSUE THREAD

**Issue by CoinCheung** [pytorch/pytorch#20683]
## 🐛 Bug

When  use gpu to train a two-convolution-layer model, the random behavior is different in the running of two times.

## To Reproduce

1. write this code in some file, on my computer the source code is named `try.py`
```python
import torch
import numpy as np
import random

torch.manual_seed(123)
random.seed(123)
np.random.seed(123)
torch.cuda.manual_seed(123)
torch.backends.cudnn.enabled=False
torch.backends.cudnn.deterministic=True

conv1 = torch.nn.Conv2d(3, 32, 3, 1, 1).cuda()
conv2 = torch.nn.Conv2d(32, 32, 3, stride=1, padding=1, dilation=1).cuda()
#  fc = torch.nn.Linear(32, 10)
criteria = torch.nn.CrossEntropyLoss()

torch.nn.init.kaiming_normal_(conv1.weight, 1)
torch.nn.init.kaiming_normal_(conv2.weight, 1)
params = list(conv1.parameters()) + list(conv2.parameters())
optim = torch.optim.SGD(params, lr=1e-4, momentum=0.9, weight_decay=1e-4)

for i in range(10):
    inten = torch.randn(16, 3, 64, 64).cuda()
    lb = torch.randint(0, 10, (16, 64, 64)).cuda()

    optim.zero_grad()
    feat = conv1(inten)
    feat = conv2(feat)
    loss = criteria(feat, lb)
    loss.backward()
    optim.step()
    print(loss.item())
```
2. run the above code for twice, and see the loss printed to the screen:
```
CUDA_VISIBLE_DEVICES=0 python try.py
3.9505960941314697
3.962101936340332
3.944667100906372
3.9629740715026855
3.9574058055877686
3.9502787590026855
3.956238031387329
3.9543356895446777
3.950388193130493
3.946221351623535

CUDA_VISIBLE_DEVICES=0 python try.py
3.9505960941314697
3.962101697921753
3.944666862487793
3.9629745483398438
3.9574055671691895
3.9502787590026855
3.956238031387329
3.9543352127075195
3.950387954711914
3.9462215900421143
```
Since my practical model has a big variance, this random behavior can bring a final score variance of as much as 0.5%, which makes it too difficult for me to tune the hyper-parameters. What is the cause of this randomness, and how could I reduce it please?

## Environment
Collecting environment information...
PyTorch version: 1.1.0a0+95ce796
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: CentOS Linux 7 (Core)
GCC version: (GCC) 5.4.0
CMake version: version 3.14.0

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 390.67
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.16.3
[pip] torch==1.1.0
[pip] torchvision==0.2.2.post3
[pip] torchvision-nightly==0.2.3
[conda] Could not collect

**Comment by colesbury** [comment]
The "2d" version of CrossEntropyLoss (nll_loss2d) uses floating-point atomicAdds and is therefore non-deterministic.

**Comment by CoinCheung** [comment]
Hi, 
But why the cpu version is doing as expected but the cuda version code does not?

**Comment by CoinCheung** [comment]
Hi, 
There is one thing that I do not quite understand, I modified my code like this: 
```python

import torch
import torch.nn.functional as F
import numpy as np
import random

torch.manual_seed(123)
random.seed(123)
np.random.seed(123)
torch.cuda.manual_seed(123)
torch.backends.cudnn.enabled = False
torch.backends.cudnn.deterministic = True

conv1 = torch.nn.Conv2d(3, 32, 3, 1, 1).cuda()
conv2 = torch.nn.Conv2d(32, 32, 3, stride=2, padding=1, dilation=1).cuda()
criteria = torch.nn.CrossEntropyLoss()

torch.nn.init.kaiming_normal_(conv1.weight, 1)
torch.nn.init.kaiming_normal_(conv2.weight, 1)
params = list(conv1.parameters()) + list(conv2.parameters())
optim = torch.optim.SGD(params, lr=1e-3, momentum=0.9, weight_decay=1e-4)


grads = {}
def save_grad(name):
    def hook(grad):
        grads[name] = torch.mean(grad).item()
    return hook

for i in range(10):
    inten = torch.randn(16, 3, 64, 64).cuda()
    lb = torch.randint(0, 10, (16, 64, 64)).cuda()

    optim.zero_grad()
    feat = conv1(inten)
    feat = conv2(feat)
    feat = F.interpolate(feat, inten.size()[2:], mode='bilinear', align_corners=True)
    N, C, H, W = feat.size()
    feat = feat.permute(0, 2, 3, 1).contiguous().view(-1, C)
    feat.register_hook(save_grad('feat'))
    lb = lb.view(-1)
    loss = criteria(feat, lb)
    loss.backward()
    optim.step()

    print(grads['feat'])
```
Since I modified the code to use 1-d cross-entropy, there should not be randomness from the cross-entropy, but the results of two running this code become:
```
-1.27675647831893e-15
-1.609823385706477e-15
-3.941291737419306e-15
-1.8318679906315083e-15
-1.1657341758564144e-15
-2.831068712794149e-15
-2.4424906541753444e-15
-2.220446049250313e-15
-3.2751579226442118e-15
-1.9984014443252818e-15
```
and 
```
-1.27675647831893e-15
-1.609823385706477e-15
-3.83026943495679e-15
-1.8318679906315083e-15
-1.1657341758564144e-15
-2.831068712794149e-15
-2.6645352591003757e-15
-1.8318679906315083e-15
-2.886579864025407e-15
-1.6653345369377348e-15
```
They are still different. Did not I eliminate the randomness by using the 1-d cross_entropy?

**Comment by gchanan** [comment]
I think this fits under https://github.com/pytorch/pytorch/issues/15359, going to mark this as a duplicate.

**Comment by sbelharbi** [comment]
Config: Python `3.7.0`. Pytorch nightly build `1.2.0.dev20190616` (https://download.pytorch.org/whl/nightly/cu100/torch_nightly-1.2.0.dev20190616-cp37-cp37m-linux_x86_64.whl)

@CoinCheung 
Hello,
I made some changes (only to shape tensors, loss).
 
From the first look, I suspected that `F.interpolate` is the reason for the non-deterministic. After inspecting, `F.interpolate` is actually the reason + `torch.nn.CrossEntropyLoss` seems to be sensitive to small variations. I already had some troubles with `F.interpolate` ([1](https://discuss.pytorch.org/t/non-deterministic-behavior-of-pytorch-upsample-interpolate/42842?u=sbelharbi), [2](https://discuss.pytorch.org/t/how-to-make-torch-nn-functional-interpolate-deterministic/42870?u=sbelharbi)), and I identified it as the source of randomness, as it is mentioned in the [documentation](https://pytorch.org/docs/stable/nn.html#torch.nn.functional.interpolate). It is the backward of `F.interpolate` that is not deterministic.

>> When using the CUDA backend, this operation may induce nondeterministic behaviour in be backward that is not easily switched off. Please see the notes on Reproducibility for background.


In your case, I recommend using `torch.nn.ConvTranspose2d` to upscale your features, since it not suitable for you to detach the interpolated features as I did in my case.

Why you didn't have this issue on CPU, but only GPU? I am not sure, but this may have something to to do with atomic operations on GPU.

I did the following comparison:
1. Code 1: use (`torch.nn.MSELoss` instead of `torch.nn.CrossEntropyLoss`) + `F.interpolate`. --> **deterministic behavior**.
2. Code 2: `torch.nn.CrossEntropyLoss` 1D + `F.interpolate`. --> **non-deterministic behavior**. You can see how the gradient started to diverge starting from iter. 3.
3. Code 3: case 1 and 2 may make us believe that the reason for non determinism in 2 is the cross-entropy. However, when removing the `F.interpolate` from 2, things seem to be deterministic. In this third case, we keep 1d cross-entropy, but we drop `F.interpolate`. --> **deterministic behavior**.

**From 1, 2, and 3, it seems that `torch.nn.CrossEntropyLoss` is sensitive to the small non-determinism of `F.interpolate`.**


# Code 1:
```python
import torch
import torch.nn.functional as F
import numpy as np
import random

torch.manual_seed(123)
random.seed(123)
np.random.seed(123)
torch.cuda.manual_seed(123)
torch.backends.cudnn.enabled = False
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.enabled = False

conv1 = torch.nn.Conv2d(3, 32, 3, 1, 1).cuda()
conv2 = torch.nn.Conv2d(32, 32, 3, stride=2, padding=1, dilation=1).cuda()
criteria = torch.nn.MSELoss()

torch.nn.init.kaiming_normal_(conv1.weight, 1)
torch.nn.init.kaiming_normal_(conv2.weight, 1)
params = list(conv1.parameters()) + list(conv2.parameters())
optim = torch.optim.SGD(params, lr=1e-3, momentum=0.9, weight_decay=1e-4)


grads = {}


def save_grad(name):
    def hook(grad):
        grads[name] = torch.mean(grad).item()
    return hook


for i in range(10):
    inten = torch.randn(16, 3, 64, 64).cuda()
    lb = torch.rand((16, 32, 64, 64)).cuda()

    optim.zero_grad()
    feat = conv1(inten)
    feat = conv2(feat)
    feat = F.interpolate(feat, inten.size()[2:], mode='bilinear', align_corners=True)
    N, C, H, W = feat.size()
    feat = feat.permute(0, 2, 3, 1).contiguous().view(-1, C)
    feat.register_hook(save_grad('feat'))
    lb = lb.view(-1, C)
    loss = criteria(feat, lb)
    loss.backward()
    optim.step()

    print("{:.20f}".format(grads['feat']))
```

# Run 1:
```bash
-0.00000045384399527393
-0.00000045347988475442
-0.00000045270792270458
-0.00000045228910039441
-0.00000045094648726263
-0.00000045106855850463
-0.00000045151733729654
-0.00000044820410494140
-0.00000044901213414050
-0.00000044841340240964
```
# Run 2:
```bash
-0.00000045384399527393
-0.00000045347988475442
-0.00000045270792270458
-0.00000045228910039441
-0.00000045094648726263
-0.00000045106855850463
-0.00000045151733729654
-0.00000044820410494140
-0.00000044901213414050
-0.00000044841340240964
```

# Code 2:
```python
import torch
import torch.nn.functional as F
import numpy as np
import random

torch.manual_seed(123)
random.seed(123)
np.random.seed(123)
torch.cuda.manual_seed(123)
torch.backends.cudnn.enabled = False
torch.backends.cudnn.deterministic = True


conv1 = torch.nn.Conv2d(3, 32, 3, 1, 1).cuda()
conv2 = torch.nn.Conv2d(32, 32, 3, stride=2, padding=1, dilation=1).cuda()
criteria = torch.nn.CrossEntropyLoss()

torch.nn.init.kaiming_normal_(conv1.weight, 1)
torch.nn.init.kaiming_normal_(conv2.weight, 1)
params = list(conv1.parameters()) + list(conv2.parameters())
optim = torch.optim.SGD(params, lr=1e-3, momentum=0.9, weight_decay=1e-4)


grads = {}


def save_grad(name):
    def hook(grad):
        grads[name] = torch.mean(grad).item()
    return hook


for i in range(10):
    inten = torch.randn(16, 3, 64, 64).cuda()
    lb = torch.randint(0, 2, (16, 64, 64)).cuda()

    optim.zero_grad()
    feat = conv1(inten)
    feat = conv2(feat)
    feat = F.interpolate(feat, inten.size()[2:], mode='bilinear', align_corners=True)
    N, C, H, W = feat.size()
    feat = feat.permute(0, 2, 3, 1).contiguous().view(-1, C)
    feat.register_hook(save_grad('feat'))
    lb = lb.view(-1)
    loss = criteria(feat, lb)
    loss.backward()
    optim.step()

    print("{:.20f}".format(grads['feat']))
```

# Run 1:
```bash
-0.00000000000000327516
-0.00000000000000199840
-0.00000000000000233147
-0.00000000000000177636
-0.00000000000000188738
0.00000000000000083267
-0.00000000000000144329
-0.00000000000000210942
-0.00000000000000166533
-0.00000000000000205391
```
# Run 2:
```bash
-0.00000000000000327516
-0.00000000000000199840
-0.00000000000000233147
-0.00000000000000166533
-0.00000000000000188738
0.00000000000000111022
-0.00000000000000199840
-0.00000000000000260902
-0.00000000000000188738
-0.00000000000000266454
```

# Code 3:
```python
import torch
import torch.nn.functional as F
import numpy as np
import random

torch.manual_seed(123)
random.seed(123)
np.random.seed(123)
torch.cuda.manual_seed(123)
torch.backends.cudnn.enabled = False
torch.backends.cudnn.deterministic = True

conv1 = torch.nn.Conv2d(3, 32, 3, 1, 1).cuda()
conv2 = torch.nn.Conv2d(32, 32, 3, stride=2, padding=1, dilation=1).cuda()
criteria = torch.nn.CrossEntropyLoss()

torch.nn.init.kaiming_normal_(conv1.weight, 1)
torch.nn.init.kaiming_normal_(conv2.weight, 1)
params = list(conv1.parameters()) + list(conv2.parameters())
optim = torch.optim.SGD(params, lr=1e-3, momentum=0.9, weight_decay=1e-4)


grads = {}


def save_grad(name):
    def hook(grad):
        grads[name] = torch.mean(grad).item()
    return hook


for i in range(10):
    inten = torch.randn(16, 3, 64, 64).cuda()
    lb = torch.randint(0, 2, (16, 32, 32)).cuda()

    optim.zero_grad()
    feat = conv1(inten)
    feat = conv2(feat)
    # feat = F.interpolate(feat, inten.size()[2:], mode='bilinear', align_corners=True)
    N, C, H, W = feat.size()
    feat = feat.permute(0, 2, 3, 1).contiguous().view(-1, C)
    feat.register_hook(save_grad('feat'))
    lb = lb.view(-1)
    loss = criteria(feat, lb)
    loss.backward()
    optim.step()

    print("{:.20f}".format(grads['feat']))
```

# Run 1:
```bash
-0.00000000000003019807
-0.00000000000000266454
-0.00000000000000976996
-0.00000000000001687539
0.00000000000000088818
-0.00000000000000444089
0.00000000000000532907
-0.00000000000001865175
-0.00000000000000444089
0.00000000000000000000
```
# Run 2:
```bash
-0.00000000000003019807
-0.00000000000000266454
-0.00000000000000976996
-0.00000000000001687539
0.00000000000000088818
-0.00000000000000444089
0.00000000000000532907
-0.00000000000001865175
-0.00000000000000444089
0.00000000000000000000

```
